File: Algorithms/Sorting/Bubble_sort/bubble_sort.cpp
#include <iostream>
using namespace std;

//this function used for swapping the array elements
int swap(int * a, int * b){
    int  temp;
    temp = *a;
    *a = *b;
    *b = temp;
    return 0;
}

//bubble sort function
void bubble_sort(int arr[], int n){
    for(int i=0; i<n-1;i++){

        int swapped = 0;
        
        for(int j=0; j<n-i-1;j++){

            if(arr[j]>arr[j+1]){ //for descending order use arr[j]<arr[j+1]
                swap(&arr[j],&arr[j+1]);
                swapped = 1;
            }
        }

        if(swapped==0){
            break;
        }
    }
}



int main(){
    int arr[] = {10,9,8,45,2,96,12}; //array size is 7
    bubble_sort(arr,7);
    for(int i=0;i<7;i++){
    	cout<<arr[i]<<" ";
	}
    return 0;
}

File: Algorithms/Sorting/Bubble_sort/bubble_sort.py
def bubble_sort(arr):
    
    for i in range(len(arr)-1):
        
        swapped = False #used for optimization
        
        for j in range(len(arr)-1-i):
            if arr[j]>arr[j+1]:
                arr[j],arr[j+1] = arr[j+1], arr[j]
                swapped = True
                
        if swapped == False:
            break
        
    print("Array after sorting is:", arr)
    

#driver code
if __name__ == '__main__':
    arr = [10,9,8,45,2,96,12]
    print("Array before sorting is ",arr)
    bubble_sort(arr)